File: apps/web/src/services/banditRecommendationService.ts
/**
 * 多臂老虎机推荐系统API服务
 */

import apiClient from './apiClient';

export interface RecommendationRequest {
  user_id: string;
  context?: Record<string, any>;
  num_recommendations?: number;
  exclude_items?: string[];
  include_explanations?: boolean;
  experiment_id?: string;
}

export interface Recommendation {
  item_id: string;
  score: number;
  confidence: number;
}

export interface RecommendationResponse {
  request_id: string;
  user_id: string;
  recommendations: Recommendation[];
  algorithm_used: string;
  confidence_score: number;
  cold_start_strategy?: string;
  explanations?: string[];
  timestamp?: string;
  processing_time_ms: number;
}

export interface FeedbackRequest {
  user_id: string;
  item_id: string;
  feedback_type: string;
  feedback_value?: number;
  context?: Record<string, any>;
}

export interface InitializeConfig {
  enable_cold_start?: boolean;
  enable_evaluation?: boolean;
}

export interface EngineStatistics {
  engine_stats: {
    total_requests: number;
    cache_hits: number;
    cold_start_requests: number;
    algorithm_usage: Record<string, number>;
    average_response_time_ms: number;
  };
  algorithm_stats: Record<string, {
    average_reward: number;
    total_reward: number;
    total_pulls: number;
    regret: number;
  }>;
  active_users: number;
  cache_size: number;
}

export interface AlgorithmInfo {
  name: string;
  display_name: string;
  description: string;
  supports_context: boolean;
  supports_binary_feedback: boolean;
}

export interface ExperimentConfig {
  experiment_name: string;
  algorithms: Record<string, {
    algorithm_type: string;
    config: Record<string, any>;
  }>;
  traffic_split: Record<string, number>;
  duration_hours: number;
  min_sample_size?: number;
}

export interface ExperimentResult {
  experiment_id: string;
  experiment_name: string;
  status: string;
  start_time: string;
  end_time: string;
  traffic_split: Record<string, number>;
  results: Record<string, {
    metrics: {
      click_through_rate: number;
      conversion_rate: number;
      average_reward: number;
      cumulative_reward: number;
      coverage: number;
      diversity: number;
    };
    sample_size: number;
    bandit_stats: Record<string, any>;
  }>;
  significance: {
    significant: boolean;
    p_value: number;
    t_statistic?: number;
    compared_variants?: string[];
  };
  recommendation: string;
}

class BanditRecommendationService {
  private readonly baseUrl = '/bandit';

  /**
   * 初始化推荐引擎
   */
  async initialize(
    nItems: number,
    config: InitializeConfig = {}
  ): Promise<{ status: string; message: string; config: any }> {
    const params = new URLSearchParams({
      n_items: nItems.toString(),
      enable_cold_start: (config.enable_cold_start ?? true).toString(),
      enable_evaluation: (config.enable_evaluation ?? true).toString(),
    });

    const response = await apiClient.post(`${this.baseUrl}/initialize?${params}`);
    return response.data;
  }

  /**
   * 获取推荐
   */
  async getRecommendations(request: RecommendationRequest): Promise<RecommendationResponse> {
    const response = await apiClient.post(`${this.baseUrl}/recommend`, request);
    return response.data;
  }

  /**
   * 提交用户反馈
   */
  async submitFeedback(feedback: FeedbackRequest): Promise<{ status: string; message: string }> {
    const response = await apiClient.post(`${this.baseUrl}/feedback`, feedback);
    return response.data;
  }

  /**
   * 获取引擎统计信息
   */
  async getStatistics(): Promise<{ status: string; statistics: EngineStatistics }> {
    const response = await apiClient.get(`${this.baseUrl}/statistics`);
    return response.data;
  }

  /**
   * 更新用户上下文
   */
  async updateUserContext(
    userId: string,
    context: Record<string, any>
  ): Promise<{ status: string; message: string }> {
    const response = await apiClient.put(`${this.baseUrl}/user/${userId}/context`, context);
    return response.data;
  }

  /**
   * 更新物品特征
   */
  async updateItemFeatures(
    itemId: string,
    features: Record<string, any>
  ): Promise<{ status: string; message: string }> {
    const response = await apiClient.put(`${this.baseUrl}/item/${itemId}/features`, features);
    return response.data;
  }

  /**
   * 获取健康状态
   */
  async getHealth(): Promise<{
    status: string;
    is_initialized: boolean;
    timestamp: string;
    engine_stats?: any;
  }> {
    const response = await apiClient.get(`${this.baseUrl}/health`);
    return response.data;
  }

  /**
   * 获取可用算法列表
   */
  async getAlgorithms(): Promise<{ status: string; algorithms: AlgorithmInfo[]; count: number }> {
    const response = await apiClient.get(`${this.baseUrl}/algorithms`);
    return response.data;
  }

  /**
   * 创建A/B测试实验
   */
  async createExperiment(experiment: ExperimentConfig): Promise<{
    status: string;
    experiment_id: string;
    experiment_name: string;
    start_time: string;
    end_time: string;
    variants: string[];
    message: string;
  }> {
    const response = await apiClient.post(`${this.baseUrl}/experiments`, experiment);
    return response.data;
  }

  /**
   * 获取A/B测试实验列表
   */
  async getExperiments(): Promise<{
    status: string;
    experiments: Array<{
      experiment_id: string;
      experiment_name: string;
      start_time: string;
      end_time: string;
      variants: string[];
      total_events: number;
    }>;
    count: number;
  }> {
    const response = await apiClient.get(`${this.baseUrl}/experiments`);
    return response.data;
  }

  /**
   * 获取A/B测试实验结果
   */
  async getExperimentResults(experimentId: string): Promise<{
    status: string;
    results: ExperimentResult;
  }> {
    const response = await apiClient.get(`${this.baseUrl}/experiments/${experimentId}/results`);
    return response.data;
  }

  /**
   * 结束A/B测试实验
   */
  async endExperiment(experimentId: string): Promise<{
    status: string;
    message: string;
    timestamp: string;
  }> {
    const response = await apiClient.post(`${this.baseUrl}/experiments/${experimentId}/end`);
    return response.data;
  }

  /**
   * 批量获取推荐（用于性能测试）
   */
  async batchRecommendations(
    requests: RecommendationRequest[]
  ): Promise<RecommendationResponse[]> {
    const promises = requests.map(request => this.getRecommendations(request));
    return Promise.all(promises);
  }

  /**
   * 多次请求推荐，收集真实返回
   */
  async simulateUserSession(
    userId: string,
    numInteractions: number,
    contextGenerator?: () => Record<string, any>
  ): Promise<{
    recommendations: RecommendationResponse[];
    feedbacks: Array<{ item_id: string; feedback_type: string; feedback_value: number }>;
    totalReward: number;
  }> {
    const recommendations: RecommendationResponse[] = [];
    const feedbacks: Array<{ item_id: string; feedback_type: string; feedback_value: number }> = [];
    let totalReward = 0;

    for (let i = 0; i < numInteractions; i++) {
      const context = contextGenerator ? contextGenerator() : undefined;
      const recResponse = await this.getRecommendations({
        user_id: userId,
        num_recommendations: 3,
        context,
        include_explanations: false
      });
      recommendations.push(recResponse);
    }

    return { recommendations, feedbacks, totalReward };
  }

  /**
   * 获取算法性能对比数据
   */
  async getAlgorithmComparison(): Promise<{
    algorithms: string[];
    metrics: {
      algorithm: string;
      average_reward: number;
      total_pulls: number;
      regret: number;
      usage_count: number;
    }[];
  }> {
    const statsResponse = await this.getStatistics();
    const stats = statsResponse.statistics;
    
    const algorithms = Object.keys(stats.algorithm_stats);
    const metrics = algorithms.map(algorithm => ({
      algorithm,
      average_reward: stats.algorithm_stats[algorithm].average_reward || 0,
      total_pulls: stats.algorithm_stats[algorithm].total_pulls || 0,
      regret: stats.algorithm_stats[algorithm].regret || 0,
      usage_count: stats.engine_stats.algorithm_usage[algorithm] || 0
    }));

    return { algorithms, metrics };
  }
}

// 导出单例实例
export const banditRecommendationService = new BanditRecommendationService();
export default banditRecommendationService;

File: apps/api/src/services/bandit_recommendation_service.py
"""
多臂老虎机推荐服务

管理推荐引擎的初始化、配置和生命周期。
"""

from typing import Dict, List, Any, Optional
from datetime import datetime
from src.core.utils.timezone_utils import utc_now
from src.ai.reinforcement_learning.recommendation_engine import (

    BanditRecommendationEngine,
    AlgorithmType,
    RecommendationRequest,
    FeedbackData
)

from src.core.logging import get_logger
logger = get_logger(__name__)

class BanditRecommendationService:
    """多臂老虎机推荐服务"""
    
    def __init__(self):
        """初始化推荐服务"""
        self.engine: Optional[BanditRecommendationEngine] = None
        self.is_initialized = False
        self.default_config = {
            "n_items": 1000,
            "default_algorithm": AlgorithmType.UCB,
            "enable_cold_start": True,
            "enable_evaluation": True,
            "cache_ttl_seconds": 300,
            "max_cache_size": 10000
        }
        self.algorithm_configs = {
            "ucb": {"c": 2.0, "random_state": 42},
            "thompson_sampling": {"alpha_init": 1.0, "beta_init": 1.0, "random_state": 42},
            "epsilon_greedy": {"epsilon": 0.1, "decay_rate": 0.995, "random_state": 42},
            "linear_contextual": {"n_features": 50, "alpha": 1.0, "lambda_reg": 1.0, "random_state": 42}
        }
        
    async def initialize(
        self,
        n_items: int = None,
        algorithm_configs: Optional[Dict[str, Dict[str, Any]]] = None,
        **kwargs
    ) -> bool:
        """
        初始化推荐引擎
        
        Args:
            n_items: 物品数量
            algorithm_configs: 算法配置
            **kwargs: 其他配置参数
            
        Returns:
            是否初始化成功
        """
        try:
            if self.is_initialized:
                logger.warning("推荐引擎已经初始化")
                return True
                
            # 合并配置
            config = self.default_config.copy()
            config.update(kwargs)
            
            if n_items:
                config["n_items"] = n_items
                
            # 创建推荐引擎
            self.engine = BanditRecommendationEngine(
                default_algorithm=config["default_algorithm"],
                enable_cold_start=config["enable_cold_start"],
                enable_evaluation=config["enable_evaluation"],
                cache_ttl_seconds=config["cache_ttl_seconds"],
                max_cache_size=config["max_cache_size"]
            )
            
            # 初始化算法
            algorithm_configs = algorithm_configs or self.algorithm_configs
            await self.engine.initialize_algorithms(config["n_items"], algorithm_configs)
            
            self.is_initialized = True
            logger.info(f"推荐引擎初始化成功，支持{config['n_items']}个物品")
            
            return True
            
        except Exception as e:
            logger.error(f"推荐引擎初始化失败: {e}")
            self.engine = None
            self.is_initialized = False
            return False
    
    async def get_recommendations(
        self,
        user_id: str,
        num_recommendations: int = 10,
        context: Optional[Dict[str, Any]] = None,
        exclude_items: Optional[List[str]] = None,
        include_explanations: bool = False,
        experiment_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        获取推荐
        
        Args:
            user_id: 用户ID
            num_recommendations: 推荐数量
            context: 上下文信息
            exclude_items: 排除的物品
            include_explanations: 是否包含解释
            experiment_id: 实验ID
            
        Returns:
            推荐结果
        """
        if not self.is_initialized or not self.engine:
            raise RuntimeError("推荐引擎未初始化")
            
        request = RecommendationRequest(
            user_id=user_id,
            context=context,
            num_recommendations=num_recommendations,
            exclude_items=exclude_items,
            include_explanations=include_explanations,
            experiment_id=experiment_id
        )
        
        response = await self.engine.get_recommendations(request)
        
        return {
            "request_id": response.request_id,
            "user_id": response.user_id,
            "recommendations": response.recommendations,
            "algorithm_used": response.algorithm_used,
            "confidence_score": response.confidence_score,
            "cold_start_strategy": response.cold_start_strategy,
            "explanations": response.explanations,
            "timestamp": response.timestamp,
            "processing_time_ms": response.processing_time_ms
        }
    
    async def process_feedback(
        self,
        user_id: str,
        item_id: str,
        feedback_type: str,
        feedback_value: float = 0.0,
        context: Optional[Dict[str, Any]] = None
    ) -> bool:
        """
        处理用户反馈
        
        Args:
            user_id: 用户ID
            item_id: 物品ID
            feedback_type: 反馈类型
            feedback_value: 反馈值
            context: 上下文信息
            
        Returns:
            是否处理成功
        """
        if not self.is_initialized or not self.engine:
            raise RuntimeError("推荐引擎未初始化")
            
        try:
            feedback = FeedbackData(
                user_id=user_id,
                item_id=item_id,
                feedback_type=feedback_type,
                feedback_value=feedback_value,
                context=context,
                timestamp=utc_now()
            )
            
            await self.engine.process_feedback(feedback)
            logger.info(f"处理用户{user_id}对物品{item_id}的反馈: {feedback_type}")
            return True
            
        except Exception as e:
            logger.error(f"处理反馈失败: {e}")
            return False
    
    async def update_user_context(self, user_id: str, context: Dict[str, Any]) -> bool:
        """
        更新用户上下文
        
        Args:
            user_id: 用户ID
            context: 上下文信息
            
        Returns:
            是否更新成功
        """
        if not self.is_initialized or not self.engine:
            raise RuntimeError("推荐引擎未初始化")
            
        try:
            await self.engine.update_user_context(user_id, context)
            return True
        except Exception as e:
            logger.error(f"更新用户上下文失败: {e}")
            return False
    
    async def update_item_features(self, item_id: str, features: Dict[str, Any]) -> bool:
        """
        更新物品特征
        
        Args:
            item_id: 物品ID
            features: 物品特征
            
        Returns:
            是否更新成功
        """
        if not self.is_initialized or not self.engine:
            raise RuntimeError("推荐引擎未初始化")
            
        try:
            await self.engine.update_item_features(item_id, features)
            return True
        except Exception as e:
            logger.error(f"更新物品特征失败: {e}")
            return False
    
    def get_statistics(self) -> Dict[str, Any]:
        """
        获取统计信息
        
        Returns:
            统计信息字典
        """
        if not self.is_initialized or not self.engine:
            raise RuntimeError("推荐引擎未初始化")
            
        try:
            return self.engine.get_engine_statistics()
        except Exception as e:
            logger.error(f"获取统计信息失败: {e}")
            return {}
    
    def get_health_status(self) -> Dict[str, Any]:
        """
        获取健康状态
        
        Returns:
            健康状态字典
        """
        status = {
            "service": "bandit_recommendation_service",
            "is_initialized": self.is_initialized,
            "timestamp": utc_now().isoformat()
        }
        
        if self.is_initialized and self.engine:
            try:
                stats = self.engine.get_engine_statistics()
                status.update({
                    "status": "healthy",
                    "engine_stats": {
                        "total_requests": stats["engine_stats"]["total_requests"],
                        "cache_hits": stats["engine_stats"]["cache_hits"],
                        "active_users": stats["active_users"],
                        "cache_size": stats["cache_size"],
                        "algorithms": list(stats["algorithm_stats"].keys())
                    }
                })
            except Exception as e:
                status.update({
                    "status": "degraded",
                    "error": str(e)
                })
        else:
            status["status"] = "not_initialized"
            
        return status
    
    async def shutdown(self):
        """关闭推荐服务"""
        if self.engine:
            # 这里可以添加保存状态、清理资源等逻辑
            logger.info("推荐引擎正在关闭")
            self.engine = None
            self.is_initialized = False

# 全局服务实例
bandit_recommendation_service = BanditRecommendationService()
